lsi/rom/tools/cp16mic.py

```diff
@@ -0,0 +1,981 @@
+                if lt > 1:
+                    if token[1] in 'xX':
+                        return int(token[2:], base=16)
+                    if token[1] in 'bB':
+                        return int(token[2:], base=2)
+                    if token[-1] == '.':
+                        return int(token[:-1], base=10)
+                    return int(token[:-1], base=8)
+            if token[-1] == '.':
+                return int(token[:-1], base=10)
+            return int(token, self.radx)
+        except ValueError:
+            raise SyntaxError("invalid integer expression '%s'" % token)
+    #
+    # Token generator wrapper, replaces the symbols with actual integers
+    #
+    def subst_symbol(self, string):
+        for token in string:
+            if token in wd16.EXP_ALL:
+                yield token
+            elif token[0] in '0123456789':
+                yield self.subst_number(token)
+            else:
+                value = self.symb.get(token)
+                if value is None:
+                    raise SyntaxError("undefined identifier '%s'" % token)
+                yield value[0]
+    #
+    # Source token generator parses the original line
+    #                 
+    def parse_exp(self, string):
+        start = -1
+        pos = -1
+        for s in string:
+            pos += 1
+            if start < 0:
+                if s in ' \t':
+                    continue
+                start = pos
+            if s.isalnum() or s in '$_.':
+                continue
+            if start >= 0:
+                yield string[start:pos]
+                start = -1
+            if s in wd16.EXP_ALL:
+                yield s
+        if start >= 0:
+            yield string[start:]
+    #
+    # Evaluate arithmetics expression, returns integer if succeeded,
+    # otherwise throws SyntaxError with error message generated 
+    #
+    def eval_exp(self, exp):
+        value = self.calc_exp(                          # calculate polish
+                    self.polish_exp(                    # convert to polish
+                        self.subst_unary(               # convert unary
+                            self.subst_symbol(          # subst symbols
+                                self.parse_exp(exp))))) # generate tokens
+        return value
+
+    def open_file(self, name, ext, mode):
+        #
+        # Add the default file name extension if needed
+        #
+        sname = os.path.splitext(name)
+        if not sname[1]:
+            name += '.' + ext
+        self.fnam = name
+        try:
+            file = open(name, mode, -1, None, None)
+        except OSError as err:
+            raise RuntimeError(err)
+        return file
+
+    def close_file(self, file):
+        if file is not None:
+            file.close()
+        return
+
+    #
+    # Show final compilation statistics
+    #
+    def final_stat(self, flst):
+        msg = '\r\nErrors: %d\r\nWarnings: %d\r\n' % (self.ecnt, self.wcnt)
```

Comment:
Assuming you have your code running on Windows and the output file is opened in "t" (text) mode, you don't need to add extra `\r` symbol there.  For Unix `\r` is not needed as well.